source/Calamari.Aws/Kubernetes/Discovery/AwsKubernetesDiscoverer.cs

```diff
@@ -0,0 +1,129 @@
+                    log.Verbose($"    Session Name: {authenticationDetails.Role.SessionName}");
+                if (authenticationDetails.Role.SessionDuration != null) 
+                    log.Verbose($"    Session Duration: {authenticationDetails.Role.SessionDuration}");
+                if (!authenticationDetails.Role.ExternalId.IsNullOrEmpty())
+                    log.Verbose($"    External Id: {authenticationDetails.Role.ExternalId}");
+            }
+            else
+            {
+                log.Verbose("  Role: No IAM Role provided.");
+            }
+
+            AWSCredentials credentials;
+            try
+            {
+                credentials = authenticationDetails.ToCredentials();
+            }
+            catch (Exception e)
+            {
+                Log.Warn("Unable to authorise credentials, see verbose log for details.");
+                Log.Verbose($"Unable to authorise credentials: {e}");
+                yield break;
+            }
+            
+            foreach (var region in authenticationDetails.Regions)
+            {
+                var client = new AmazonEKSClient(credentials,
+                    RegionEndpoint.GetBySystemName(region));
+
+                var clusters = client.ListClustersAsync(new ListClustersRequest()).GetAwaiter().GetResult();
+
+                foreach (var cluster in clusters.Clusters.Select(c =>
+                    client.DescribeClusterAsync(new DescribeClusterRequest { Name = c }).GetAwaiter().GetResult().Cluster))
+                {
+                    var credentialsRole = authenticationDetails.Role;
+                    var assumedRole = credentialsRole.Type == "assumeRole"
+                        ? new AwsAssumeRole(credentialsRole.Arn,
+                            credentialsRole.SessionName,
+                            credentialsRole.SessionDuration,
+                            credentialsRole.ExternalId)
+                        : null;
+                    
+                    yield return KubernetesCluster.CreateForEks(cluster.Arn,
+                        cluster.Name,
+                        cluster.Endpoint,
+                        authenticationDetails.Credentials.AccountId,
+                        assumedRole,
+                        discoveryContext.Scope.WorkerPoolId,
+                        cluster.Tags.ToTargetTags());
+                }
+            }
+        }
+        
+        bool TryGetDiscoveryContext(string json, 
+            out TargetDiscoveryContext<AwsAuthenticationDetails> discoveryContext)
+        {
+            discoveryContext = null;
+            try
+            {
+                discoveryContext =
+                    JsonConvert.DeserializeObject<TargetDiscoveryContext<AwsAuthenticationDetails>>(json);
+                
+                if (discoveryContext != null)
+                    return true;
+                
+                log.Warn("Target discovery context is in the wrong format.");
```

Comment:
I think these warnings need changing too.
The way it is handled above looks good, in that there is a warning that lets the user know it is not something they can solve themselves, and need to contact Octopus support, and additional details in the verbose log.
The same approach should be taken here and in the catch statement below.
All three places could call a single method to write the log output to guaranteed consistency.

The same approach should be followed in the Azure discovery code too.

Reply:
Fixed - I ended up created a `KubernetesDiscovererBase` which implements the basic json parsing code and error logging. This keeps everything in one place and works for both Azure and AWS 👍